sdk/python/runtime/main.go

```diff
@@ -24,44 +35,57 @@ const (
 //go:embed scripts/runtime.py
 var runtimeTmpl string
 
-func (m *PythonSdk) ModuleRuntime(modSource *Directory, subPath string, introspectionJson string) *Container {
-	return m.CodegenBase(modSource, subPath, introspectionJson).
+func (m *PythonSdk) ModuleRuntime(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
+	return ctr.
 		WithExec([]string{"python", "-m", "pip", "install", "."}).
-		WithWorkdir(ModSourceDirPath).
 		WithNewFile(RuntimeExecutablePath, ContainerWithNewFileOpts{
 			Contents:    runtimeTmpl,
 			Permissions: 0755,
 		}).
-		WithEntrypoint([]string{RuntimeExecutablePath})
+		WithEntrypoint([]string{RuntimeExecutablePath}), nil
 }
 
-func (m *PythonSdk) Codegen(modSource *Directory, subPath string, introspectionJson string) *GeneratedCode {
-	ctr := m.CodegenBase(modSource, subPath, introspectionJson)
+func (m *PythonSdk) Codegen(ctx context.Context, modSource *ModuleSource, introspectionJson string) (*GeneratedCode, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
 	ctr = ctr.WithDirectory(genDir, ctr.Directory(sdkSrc), ContainerWithDirectoryOpts{
 		Exclude: []string{
 			"**/__pycache__",
 		},
 	})
 
 	modified := ctr.Directory(ModSourceDirPath)
-	diff := modSource.Diff(modified).Directory(subPath)
+	diff := modSource.RootDirectory().Diff(modified)
 
 	return dag.GeneratedCode(diff).
-		WithVCSIgnoredPaths([]string{
-			genDir,
-		})
+		WithVCSGeneratedPaths([]string{
+			genDir + "/**",
+		}), nil
 }
 
-func (m *PythonSdk) CodegenBase(modSource *Directory, subPath string, introspectionJson string) *Container {
+func (m *PythonSdk) CodegenBase(ctx context.Context, modSource *ModuleSource, introspectionJson string) (*Container, error) {
+	subPath, err := modSource.Subpath(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %v", err)
+	}
+
 	return m.Base("").
-		WithDirectory(sdkSrc, dag.Host().Directory(root(), HostDirectoryOpts{
-			Exclude: []string{"runtime"},
-		})).
-		WithMountedDirectory("/opt", dag.Host().Directory(root(), HostDirectoryOpts{
-			Include: []string{"runtime/template"},
-		})).
+		WithMountedDirectory(sdkSrc, m.SDKSourceDir).
```

Comment:
Does this include `./runtime`? Need to `Exclude` it with `WithDirectory` instead, no?

Reply:
For now, just updated it to exclude `./runtime` here via `m.SDKSourceDir.WithoutDirectory("runtime")`. Due to the way that the SDKs are loaded into the engine right now, it'd actually require the engine do more work to separately import with an exclude and provided that as `m.SDKSourceDir`. `WithoutDirectory` is actually faster here and avoids a little more special casing in the engine.

Reply:
Good to know! Regretfully I'm most likely to not remember this when it becomes useful. 😅